subsys/usb/device_next/usbd_ch9.c

```diff
@@ -0,0 +1,871 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <zephyr/kernel.h>
+#include <zephyr/usb/usbd.h>
+#include <zephyr/drivers/usb/udc.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/slist.h>
+
+#include "usbd_device.h"
+#include "usbd_desc.h"
+#include "usbd_ch9.h"
+#include "usbd_config.h"
+#include "usbd_class.h"
+#include "usbd_class_api.h"
+#include "usbd_interface.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(usbd_ch9, CONFIG_USBD_LOG_LEVEL);
+
+#define CTRL_AWAIT_SETUP_DATA		0
+#define CTRL_AWAIT_STATUS_STAGE		1
+
+static void ch9_set_ctrl_type(struct usbd_contex *const uds_ctx,
+				   const int type)
+{
+	uds_ctx->ch9_data.ctrl_type = type;
+}
+
+static int ch9_get_ctrl_type(struct usbd_contex *const uds_ctx)
+{
+	return uds_ctx->ch9_data.ctrl_type;
+}
+
+static void ctrl_set_req_compl(struct usbd_contex *const uds_ctx,
+			       void *const cb, void *const data)
+{
+	uds_ctx->ch9_data.req_compl.cb = cb;
+	uds_ctx->ch9_data.req_compl.data = data;
+}
+
+static int sreq_set_address_cb(struct usbd_contex *const uds_ctx,
+			       void *const data)
+{
+	struct usb_setup_packet *setup = usbd_get_setup_pkt(uds_ctx);
+	int ret;
+
+	ARG_UNUSED(data);
+	ret = udc_set_address(uds_ctx->dev, setup->wValue);
+	if (ret) {
+		LOG_ERR("Failed to set device address %d", setup->wValue);
+		return ret;
+	}
+
+	if (usbd_state_is_address(uds_ctx) && setup->wValue == 0) {
+		uds_ctx->ch9_data.state = USBD_STATE_DEFAULT;
+	} else {
+		uds_ctx->ch9_data.state = USBD_STATE_ADDRESS;
+	}
+
+	LOG_DBG("Set new address %d", setup->wValue);
+
+	return ret;
+}
+
+static int sreq_set_address(struct usbd_contex *const uds_ctx)
+{
+	struct usb_setup_packet *setup = usbd_get_setup_pkt(uds_ctx);
+
+	/* Not specified if wLength is non-zero, treat as error */
+	if (setup->wValue > 127 || setup->wLength) {
+		errno = -ENOTSUP;
+		return 0;
+	}
+
+	if (setup->RequestType.recipient != USB_REQTYPE_RECIPIENT_DEVICE) {
+		errno = -ENOTSUP;
+		return 0;
+	}
+
+	if (usbd_state_is_configured(uds_ctx)) {
+		errno = -EPERM;
+		return 0;
+	}
+
+	ctrl_set_req_compl(uds_ctx, sreq_set_address_cb, NULL);
```

Comment:
Having a generic callback mechanism for the **ONLY** request that works differently seems like an overkill to me. The `SET ADDRESS` should be handled separately as stated in Universal Serial Bus Specification Revision 2.0, 9.4.6 Set Address
> The USB device does not change its device address until after the Status stage of this request is completed successfully. Note that this is a difference between this request and all other requests. For all other requests, the operation indicated must be completed before the Status stage.


The reason why status stage is there is to allow host know that whole processing is done. The device firmware should not even attempt to get a notification that the status stage is finished as it is completely unnecessary (the **ONLY** exception is `SET ADDRESS`) from device perspective. When the device allows status stage to proceed, it means that the processing from device perspective is over and no further action will be taken regarding the request.

Note that if there is some long action after the control transfer, like for example flashing after `DFU DNLD`, it is up to the class to provide polling mechanism for host to determine when the action has finished (DFU does so with `GETSTATUS`).